libvast/vast/system/format_factory.hpp

```diff
@@ -0,0 +1,138 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#ifndef FORMAT_FACTORY_HPP 
+#define FORMAT_FACTORY_HPP
+
+#include <string>
+#include <functional>
+
+#include <caf/local_actor.hpp>
+
+#include "vast/system/source.hpp"
+//#include "vast/system/spawn.hpp"
+
+#include "vast/format/bgpdump.hpp"
+#include "vast/format/mrt.hpp"
+#include "vast/format/bro.hpp"
+#ifdef VAST_HAVE_PCAP
+#include "vast/format/pcap.hpp"
+#endif
+
+#include "vast/detail/make_io_stream.hpp"
+
+namespace vast::system {
+
+class format_factory {
+public:
+  // TODO: document me
+  template <class Format>
+  using format_factory_function = std::function<expected<Format>(caf::message&)>;
+
+  using actor_factory_function
+    = std::function<caf::actor(caf::local_actor*, caf::message&)>;
+
+  struct default_args {
+    default_args(caf::message& args) 
+      : args(args) {
+      parse_result = args.extract_opts(
+        {{"read,r", "path to input where to read events from", input},
+         {"schema,s", "path to alternate schema", schema_file},
+         {"uds,d", "treat -r as listening UNIX domain socket", uds}});
+    }
+
+    ~default_args() {
+      // Ensure that, we update the parameter list such that it no longer
+      // contains the command line options that we have used
+      args = parse_result.remainder;
+    }
+
+    caf::message& args;
+    std::string input = "-"s;
+    std::string schema_file;
+    bool uds = false;
+    caf::message::cli_res parse_result;
+  };
```

Comment:
This struct contains two orthogonal concerns that should be disentangled:
1. Default arguments
2. Modification of the command line